### ISSUE TITLE
OOM when creating custom phrase vectors.

### ISSUE THREAD

**Issue by greenhornbb** [princeton-nlp/DensePhrases#36]
### Description
DensePhrases has been a significant influence on my current project, and I am working on creating a custom index based on the enwiki-20240920-dump.

I have successfully generated the articles.json file, which is around 18GB in size. However, when attempting to create the phrase vectors using commands similar to those in the instructions, I have encountered persistent OOM (Out of Memory) issues. I also tried utilizing distributed training to resolve the issue, but it seems to worsen performance rather than improving it.

### Commands
CUDA_VISIBLE_DEVICES=1,2,3,4 TORCH_DISTRIBUTED_DEBUG=DETAIL python -m torch.distributed.launch --nproc_per_node=4 --rdzv_endpoint=127.0.0.1:29500  generate_phrase_vecs.py \
    --model_type bert \
    --pretrained_name_or_path SpanBERT/spanbert-base-cased \
    --data_dir ./ \
    --cache_dir $CACHE_DIR \
    --predict_file examples/create-custom-index/combined_output01.json \
    --do_dump \
    --max_seq_length 512 \
    --doc_stride 500 \
    --fp16 \
    --filter_threshold -2.0 \
    --append_title \
    --load_dir $SAVE_DIR/densephrases-multi \
    --output_dir $SAVE_DIR/densephrases-multi_sample
    
    ### Logging
    10/31/2024 15:02:21 - INFO - transformers.tokenization_utils -   Model name 'SpanBERT/spanbert-base-cased' not found in model shortcut name list (bert-base-uncased, bert-large-uncased, bert-base-cased, bert-large-cased, bert-base-multilingual-uncased, bert-base-multilingual-cased, bert-base-chinese, bert-base-german-cased, bert-large-uncased-whole-word-masking, bert-large-cased-whole-word-masking, bert-large-uncased-whole-word-masking-finetuned-squad, bert-large-cased-whole-word-masking-finetuned-squad, bert-base-cased-finetuned-mrpc, bert-base-german-dbmdz-cased, bert-base-german-dbmdz-uncased, bert-base-finnish-cased-v1, bert-base-finnish-uncased-v1, bert-base-dutch-cased). Assuming 'SpanBERT/spanbert-base-cased' is a path, a model identifier, or url to a directory containing tokenizer files.
10/31/2024 15:02:25 - INFO - transformers.tokenization_utils -   loading file https://s3.amazonaws.com/models.huggingface.co/bert/SpanBERT/spanbert-base-cased/vocab.txt from cache at .//cache/450d2ca167544c8b0b1363bd390701766966fb14a0a03531c76a7e8ea8231d6b.e13dbb970cb325137104fb2e5f36fe865f27746c6b526f6352861b1980eb80b1
10/31/2024 15:02:25 - INFO - transformers.tokenization_utils -   loading file https://s3.amazonaws.com/models.huggingface.co/bert/SpanBERT/spanbert-base-cased/added_tokens.json from cache at None
10/31/2024 15:02:25 - INFO - transformers.tokenization_utils -   loading file https://s3.amazonaws.com/models.huggingface.co/bert/SpanBERT/spanbert-base-cased/special_tokens_map.json from cache at None
10/31/2024 15:02:25 - INFO - transformers.tokenization_utils -   loading file https://s3.amazonaws.com/models.huggingface.co/bert/SpanBERT/spanbert-base-cased/tokenizer_config.json from cache at None
10/31/2024 15:02:40 - INFO - __main__ -   Dump parameters Namespace(adam_epsilon=1e-08, append_title=True, cache_dir='.//cache', config_name='', data_dir='./', dense_offset=-2, dense_scale=20, device=device(type='cuda', index=0), do_dump=True, do_eval=False, do_filter_test=False, do_lower_case=False, do_train=False, doc_stride=500, draft=False, eval_all_checkpoints=False, evaluate_during_training=False, filter_only=False, filter_threshold=-2.0, filter_threshold_list='-1e8', fp16=True, fp16_opt_level='O1', gradient_accumulation_steps=1, lambda_flt=0.0, lambda_kl=0.0, lambda_neg=0.0, lang_id=0, learning_rate=3e-05, load_dir='.//outputs/densephrases-multi', local_rank=0, logging_steps=5000, max_answer_length=10, max_grad_norm=1.0, max_query_length=64, max_seq_length=512, max_steps=-1, model_type='bert', n_best_size=20, n_gpu=1, no_cuda=False, null_score_diff_threshold=0.0, num_train_epochs=2.0, output_dir='.//outputs/densephrases-multi_sample', overwrite_output_dir=False, pbn_size=0, pbn_tolerance=9999, per_gpu_eval_batch_size=12, per_gpu_train_batch_size=12, predict_file='examples/create-custom-index/combined_output01.json', pretrained_name_or_path='SpanBERT/spanbert-base-cased', save_steps=9999999, seed=42, teacher_dir=None, threads=20, tokenizer_name='', train_file=None, truecase=False, truecase_path='truecase/english_with_questions.dist', verbose_logging=False, version_2_with_negative=False, wandb=False, warmup_steps=0, weight_decay=0.0)
10/31/2024 15:02:40 - INFO - transformers.configuration_utils -   loading configuration file https://s3.amazonaws.com/models.huggingface.co/bert/SpanBERT/spanbert-base-cased/config.json from cache at .//cache/f657841af52c33ad17850f7918b10fbfc48e447d49d35bee4081df30d7b54545.e736d0f2e9459c34485cfbdd4c15e2b18d74c1c4359ee99166a419eaaab2994b
10/31/2024 15:02:40 - INFO - transformers.configuration_utils -   Model config BertConfig {
  "attention_probs_dropout_prob": 0.1,
  "directionality": "bidi",
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "type_vocab_size": 2,
  "vocab_size": 28996
}

10/31/2024 15:02:41 - INFO - transformers.configuration_utils -   loading configuration file https://s3.amazonaws.com/models.huggingface.co/bert/SpanBERT/spanbert-base-cased/config.json from cache at .//cache/f657841af52c33ad17850f7918b10fbfc48e447d49d35bee4081df30d7b54545.e736d0f2e9459c34485cfbdd4c15e2b18d74c1c4359ee99166a419eaaab2994b
10/31/2024 15:02:41 - INFO - transformers.configuration_utils -   Model config BertConfig {
  "attention_probs_dropout_prob": 0.1,
  "directionality": "bidi",
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "type_vocab_size": 2,
  "vocab_size": 28996
}

10/31/2024 15:02:41 - INFO - transformers.tokenization_utils -   Model name 'SpanBERT/spanbert-base-cased' not found in model shortcut name list (bert-base-uncased, bert-large-uncased, bert-base-cased, bert-large-cased, bert-base-multilingual-uncased, bert-base-multilingual-cased, bert-base-chinese, bert-base-german-cased, bert-large-uncased-whole-word-masking, bert-large-cased-whole-word-masking, bert-large-uncased-whole-word-masking-finetuned-squad, bert-large-cased-whole-word-masking-finetuned-squad, bert-base-cased-finetuned-mrpc, bert-base-german-dbmdz-cased, bert-base-german-dbmdz-uncased, bert-base-finnish-cased-v1, bert-base-finnish-uncased-v1, bert-base-dutch-cased). Assuming 'SpanBERT/spanbert-base-cased' is a path, a model identifier, or url to a directory containing tokenizer files.
10/31/2024 15:02:45 - INFO - transformers.tokenization_utils -   loading file https://s3.amazonaws.com/models.huggingface.co/bert/SpanBERT/spanbert-base-cased/vocab.txt from cache at .//cache/450d2ca167544c8b0b1363bd390701766966fb14a0a03531c76a7e8ea8231d6b.e13dbb970cb325137104fb2e5f36fe865f27746c6b526f6352861b1980eb80b1
10/31/2024 15:02:45 - INFO - transformers.tokenization_utils -   loading file https://s3.amazonaws.com/models.huggingface.co/bert/SpanBERT/spanbert-base-cased/added_tokens.json from cache at None
10/31/2024 15:02:45 - INFO - transformers.tokenization_utils -   loading file https://s3.amazonaws.com/models.huggingface.co/bert/SpanBERT/spanbert-base-cased/special_tokens_map.json from cache at None
10/31/2024 15:02:45 - INFO - transformers.tokenization_utils -   loading file https://s3.amazonaws.com/models.huggingface.co/bert/SpanBERT/spanbert-base-cased/tokenizer_config.json from cache at None
10/31/2024 15:02:45 - INFO - densephrases.utils.single_utils -   DensePhrases encoder loaded from .//outputs/densephrases-multi
10/31/2024 15:02:45 - INFO - transformers.modeling_utils -   loading weights file .//outputs/densephrases-multi/pytorch_model.bin
10/31/2024 15:02:45 - INFO - densephrases.encoder -   Loading encoders from load_dir
10/31/2024 15:03:07 - INFO - densephrases.utils.single_utils -   Load only phrase encoders for embedding phrases
10/31/2024 15:03:08 - INFO - densephrases.utils.single_utils -   Number of model parameters: 108,311,810
10/31/2024 15:03:08 - INFO - __main__ -   ***** Pre-processing contexts from examples/create-custom-index/combined_output01.json *****
10/31/2024 15:03:08 - INFO - densephrases.utils.squad_utils -   Creating features from dataset file at ./

0it [00:00, ?it/s]
6it [00:00, 42.25it/s]
18it [00:00, 72.69it/s]
28it [00:00, 71.84it/s]
36it [00:00, 65.78it/s]

### System Info
- OS: Ubuntu 20.04
- CUDA: 11.3
- Model: DensePhrases

Hi @jhyuklee , I was hoping you could provide some guidance on the OOM issues I encountered.
